driver/SDL/sdl_common.c

```diff
@@ -186,25 +182,31 @@ void mousewheel_handler(SDL_Event * event)
  */
 void keyboard_handler(SDL_Event * event)
 {
+    bool key_state = (event->type == SDL_KEYUP) ? LV_INDEV_STATE_RELEASED : LV_INDEV_STATE_PRESSED;
     /* We only care about SDL_KEYDOWN and SDL_TEXTINPUT events */
     switch(event->type) {
         case SDL_KEYDOWN:                       /*Button press*/
+        case SDL_KEYUP:                       /*Button release*/
         {
             const uint32_t ctrl_key = keycode_to_ctrl_key(event->key.keysym.sym);
-            if (ctrl_key == '\0')
-                return;
-            const size_t len = strlen(buf);
-            if (len < KEYBOARD_BUFFER_SIZE - 1) {
-                buf[len] = ctrl_key;
-                buf[len + 1] = '\0';
+            
+            if(key_buffer_len < KEYBOARD_BUFFER_SIZE){
+                key_buffer[key_buffer_len].key = ctrl_key;
+                key_buffer[key_buffer_len].state = key_state;
+                key_buffer_len++;
             }
             break;
         }
         case SDL_TEXTINPUT:                     /*Text input*/
         {
-            const size_t len = strlen(buf) + strlen(event->text.text);
-            if (len < KEYBOARD_BUFFER_SIZE - 1)
-                strcat(buf, event->text.text);
+            const size_t len = key_buffer_len + strlen(event->text.text);
+            if (len < KEYBOARD_BUFFER_SIZE - 1) {
+                for (uint32_t i = 0; i < strlen(event->text.text); i++) {
+                    key_buffer[key_buffer_len + i].key = event->text.text[i];
+                    key_buffer[key_buffer_len + i].state = key_state;
```

Comment:
Is it correct to set `state = key_state` here?  
Since we are under  `SDL_TEXTINPUT`, key_state will always be "PRESSED" here.
So when is each of these keys "released"?  

Perhaps you could add two events for each character, "press" and "release", to simulate a click on a key.

Reply:
Here is a good explanation of what SDL_TEXTINPUT does: https://wiki.libsdl.org/Tutorials-TextInput

So what would happen is that you would get events for special characters. Defaulting to PRESSED is okay as the implementation before did the same I think.

Reply:
> So what would happen is that you would get events for special characters. Defaulting to PRESSED is okay as the implementation before did the same I think.

What previous implementation did, was to send both "PRESSED" and "RELEASED" states for each  character, one after the other. This simulates a key press where a key was pressed and immediately released and that is the reason you weren't able to detect situations where two keys were pressed simultaneously.  

But here you are sending only "PRESSED" and not "RELEASED" event. That's why I suggested adding two events for each character, a "press" event followed by a "release" event, to simulate a key click.

> Here is a good explanation of what SDL_TEXTINPUT does: https://wiki.libsdl.org/Tutorials-TextInput

Thanks for the link, I wasn't familiar with these subtleties.   

So suppose the user enables IME and sends some characters to the application.  
Don't you think we need both "PRESSED" and "RELEASED" events for each character? Otherwise, we might get the impression that all these characters were pressed together and never released.

Reply:
Okay, I have added a released event directly afterwards. 

In my test application I can now see the default LVGL events when they are being hold but I still can't reproduce an LV_EVENT_LONG_PRESSED_REPEAT with pressing and holding enter. Is this intended? The behavior changes also when I add a group and focus it. Then I receive a pressed and released event for every key everytime.

Also the not-defined keys (SDL_KEYS) are now showing a released event immediately after being pressed. I am not sure what is happening here. This is my testing program which I run locally: https://sim.lvgl.io/v9.0/micropython/ports/javascript/index.html?script_direct=f1e7b7008363a47c9837289fe961a0bf8412143b

Reply:
I tried to run your code and I can see multiple problems.  

It looks like SDL is sending **both** `SDL_KEY*` and `SDL_TEXTINPUT` events for **the same** click.  
This, or course, is a source of confusion, since LVGL sees every key twice.

I've tried removing `SDL_TEXTINPUT` altogether, although this is incorrect not only for IME but also for simple <kbd>SHIFT</kbd>+key (for typing capital letters or punctuation characters for example).

Next problem was that LVGL [sets `key` in `lv_indev_data_t` to the last key automatically](https://github.com/lvgl/lvgl/blob/46cbbd672f8cf95323c2706b40698c337899004c/src/hal/lv_hal_indev.c#L176-L178), but only the last key that was **released**. So it looks like LVGL itself was not built to handle correctly simultaneous key presses.  

To work around this I tried to always set `data` in `sdl_keyboard_read` to either `key_buffer[0]` or, when `key_buffer` is empty, to the last state and key that were set, like this:

```diff
diff --git a/driver/SDL/sdl_common.c b/driver/SDL/sdl_common.c
index 071a1f1..4ef8511 100644
--- a/driver/SDL/sdl_common.c
+++ b/driver/SDL/sdl_common.c
@@ -79,14 +79,16 @@ void sdl_mousewheel_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
 void sdl_keyboard_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
 {
     (void) indev_drv;      /*Unused*/
+    static struct key_event last_state = {0};
 
     if(key_buffer_len > 0){
-        data->state = key_buffer[0].state;
-        data->key = key_buffer[0].key;
+        last_state = key_buffer[0];
         key_buffer_len--;
         data->continue_reading = key_buffer_len > 0;
         memmove(&key_buffer[0], &key_buffer[1], sizeof(struct key_event) * key_buffer_len);
     }
+    data->state = last_state.state;
+    data->key = last_state.key;
 }
```

With this fix, the test program prints the "Pressed" and "Released" events correctly, even in cases where multiple keys are pressed simultaneously.  
However, LVGL does not behave correctly in such cases and loses key events...

So to summarize, the challenges I see are:
- How to handle both `SDL_KEY*` and `SDL_TEXTINPUT` events without sending duplicated key events to LVGL
- How to fix LVGL to work correctly in situations where keys are pressed simultaneously (@kisvegabor and @embeddedt might have some ideas about that)

Reply:
> How to fix LVGL to work correctly in situations where keys are pressed simultaneously (@kisvegabor and @embeddedt might have some ideas about that)

`SDL_TEXTINPUT` is highly recommended as it handles all the these tricky cases. `SDL_KEYDOWN` should be used only to simulate embedded HW buttons in a very simple way.

Reply:
@kisvegabor - One disadvantage of SDL_TEXTINPUT is that it doesn't detect simultaneous key presses, which is the motivation for this change.   
I suspect that regardless of the specific driver implementation, LVGL doesn't support simultaneous key presses (where multiple keys might be pressed before released), because when the driver send multiple STATE_PR before sending the corresponding STATE_REL, characters are missed during typing.

Reply:
> One disadvantage of SDL_TEXTINPUT is that it doesn't detect simultaneous key presses, which is the motivation for this change.

What do you exactly mean by "simultaneous key presses"? E.g. when If press `S` and `D` on my notebook at the same time I get: sddddddddddddddddddd 

So what would be expected behavior?

Reply:
The expected behavior will be not to lose key strokes....
Here is an example:  

https://sim.lvgl.io/v9.0/micropython/ports/javascript/index.html?script_direct=1f683e2db0e13bb25905bbfeaec508d5924b9bd4

The first line in the textarea is `ABC` where each key is pressed and released in turn.
The second and third lines are the result of the same `ABC` key presses, where their release is in different order.  
As you can see, on the second and third lines, only `A` is displayed by LVGL.

Reply:
In the online C examples `SDL_TEXTINPUT` is enabled and text areas  work like any text input. You can try it [here](https://docs.lvgl.io/master/widgets/textarea.html#simple-text-area). I don't know exactly how the current driver in MP works but I think it's a reasonable default to be similar to other text inputs on desktop/web.

I think the root of the problem is that PC keyboard sends the keys on press while LVGL sends them on release. If we send the keys on release, the newly pressed key will overwrite the old one, and the last one will be released. 

We a little trick the read_cb can be modified to react on press: https://sim.lvgl.io/v9.0/micropython/ports/javascript/index.html?script_direct=efcd5f334fb6b46a367f591d77acb3088c435061

Reply:
> We a little trick the read_cb can be modified to react on press:

@kisvegabor In that trick you triggered a fake "release" for every "press".   
If I understand correctly, you are saying that in order for a driver to work correctly, every "press" must be immediately followed by a "release" of the same key....

So now I don't understand the point in `data.state`.  
If every "press" **must** be followed by a "release" (or we lose key strokes), why can't we give up `data.state` altogether?  
In each call `read_cb` could update only `data.key`, and let LVGL do the rest. Why do we need to bother with "press" and "release" states, unless they have some meaning?

Reply:
Without letting LVGL to handle the pressed and released state LVGL couldn't manage the long pressed and long pressed repeat events. So it's something for something. 

The simplest would be to make LVGL send the key events on press instead of release. I have to try it out as it might have unexpected consequences but seems working at first glance.